File: src/services/providerService.ts
import api from './api';
import { ApiResponse } from '@/types';
import { 
  ProviderProfileResponse, 
  ProviderRegisterRequest, 
  ProviderUpdateRequest 
} from '@/types/provider';

export const providerService = {
  // Provider self-service
  registerAsProvider: async (data: ProviderRegisterRequest) => {
    const response = await api.post<ApiResponse<ProviderProfileResponse>>('/provider/register', data);
    return response.data;
  },

  getMyProfile: async () => {
    const response = await api.get<ApiResponse<ProviderProfileResponse>>('/provider/me');
    return response.data;
  },

  updateMyProfile: async (data: ProviderUpdateRequest) => {
    const response = await api.put<ApiResponse<ProviderProfileResponse>>('/provider/me', data);
    return response.data;
  },

  // Public
  getPublicProfile: async (id: number) => {
    const response = await api.get<ApiResponse<ProviderProfileResponse>>(`/provider/${id}`);
    return response.data;
  }
};

File: src/main/java/com/travel/marketplace/modules/provider/service/ProviderService.java
package com.travel.marketplace.modules.provider.service;

import com.travel.marketplace.modules.provider.dto.ProviderProfileResponse;
import com.travel.marketplace.modules.provider.dto.ProviderRegisterRequest;
import com.travel.marketplace.modules.provider.dto.ProviderUpdateRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Business contract for Provider profile management.
 */
public interface ProviderService {

    /** Registers a customer as a provider. Creates a ProviderProfile and assigns the provider role. */
    ProviderProfileResponse registerAsProvider(String userEmail, ProviderRegisterRequest request);

    /** Returns the provider profile for the currently authenticated provider. */
    ProviderProfileResponse getMyProfile(String userEmail);

    /** Updates the profile for the currently authenticated provider. */
    ProviderProfileResponse updateMyProfile(String userEmail, ProviderUpdateRequest request);

    /** Returns a public (sanitized) view of a provider profile by ID. */
    ProviderProfileResponse getPublicProfile(Long providerId);

    // ── Admin operations ───────────────────────────────────────────

    /** Returns a paginated list of all provider profiles (Admin only). */
    Page<ProviderProfileResponse> getAllProviders(Pageable pageable);

    /** Approves a provider application (Admin only). */
    ProviderProfileResponse approveProvider(Long providerId);

    /** Rejects a provider application with a reason (Admin only). */
    ProviderProfileResponse rejectProvider(Long providerId, String reason);

    /** Suspends an approved provider (Admin only). */
    ProviderProfileResponse suspendProvider(Long providerId, String reason);
}
